### PR TITLE
CRUD Api for products

### PR DESCRIPTION
**Pull request by shyam-dev10**

### CODE REVIEW

File: server/models/product.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import mongoose from 'mongoose';
+import { v4 as uuidv4 } from 'uuid';
+
+const productSchema = new mongoose.Schema({
+  id: {
+    type: String,
+    default: () => uuidv4(), // You need to import the uuidv4 function
```
Comment (pavankpdev) [inline server/models/product.js]: This can be simplified with just `default: uuidv4`.
Also remove the comment

**Review by pavankpdev — CHANGES_REQUESTED** [review]
Create a single bruno file, and maintain all the request record inside that, please modify it

File: server/routes/product.js
Comment (pavankpdev) [inline server/routes/product.js:1]: Seems like input validation was skipped, I tried to provide incorrect types for the value, for example, I used a number for the properties that were suppose to be string and the API returned success status

image

This should be fixed.

**Review by pavankpdev — APPROVED** [review]
1. Update `.example.env` with all the new ENVs created, keep it up to date in the future as well. 
2. Fix the issues reported

**Review by pavankpdev — CHANGES_REQUESTED** [review]
Add input validations to API. This is a mandatory step, every single data that the API needs as an input should be validated.

**Review by pavan38kp — COMMENTED** [review]
@Shyam-crypto Seems better now, I'll run it from my end before approving the PR

**Review by pavankpdev — COMMENTED** [review]
Looks good, I'll test it form my end once, before approving

**Comment by shyam-dev10** [comment]
Okay👍

**Comment by pavankpdev** [comment]
Hey, @Shyam-crypto seems like there's a bug, I've been stuck here, no matter what type data I pass, I get the same response. 

Request
`POST /api/products/products`

image

**Comment by pavankpdev** [comment]
I passed some random string to intervals, it went through, there should be a check regarding a certain type of string the interval should accept right? 

such as `monthly` `annually`

Request 
`PUT /api/products/update/:id`

image

**Comment by pavankpdev** [comment]
One more thing, the `getAllProducts` and `getProductByID` are private routes why? 
I said that manually check for the access token in the headers, if exist show the edit mode else return only read mode. 

How can we expect users to get the product, if it's private

**Comment by shyam-dev10** [comment]
Okay i will correct it.

**Comment by pavankpdev** [comment]
@Shyam-crypto about the other two issues 

1: https://github.com/Punithshetty321/subscriberly/pull/3#issuecomment-1863217240
2: https://github.com/Punithshetty321/subscriberly/pull/3#issuecomment-1863210489

**Comment by shyam-dev10** [comment]
@pavankpdev updated the getProductById and corrected the error from interval.

File: server/controller/product.js [hunk 2]
```diff
@@ -89,43 +83,51 @@ const deleteProduct = async (req, res) => {
...
 const getProductById = async (req, res) => {
   try {
     const { id } = req.params;
-    const userId = req.user.userId;
 
-    const product = await Product.findById({ _id: id});
+    const token = req.header('Authorization');
+    let mode = 'read-only';
+
+    if (token) {
```
Comment (pavankpdev) [inline server/controller/product.js:94]: noice

**Review by pavankpdev — COMMENTED** [review]
Looks good, I'll test the updates before merging, hope bruno files are updated

**Comment by pavankpdev** [comment]
Everything looks good, merging the PR

**Pull request merged**